hash_envelope.go

```diff
@@ -0,0 +1,209 @@
+//
+// # Experimental
+//
+// Notice: The COSE Hash Envelope API is EXPERIMENTAL and may be changed or
+// removed in a later release.
+func VerifyHashEnvelope(verifier Verifier, envelope []byte) (*Sign1Message, error) {
+	// parse and validate the Hash_Envelope object
+	var message Sign1Message
+	if err := message.UnmarshalCBOR(envelope); err != nil {
+		return nil, err
+	}
+	if err := validateHashEnvelopeHeaders(&message.Headers); err != nil {
+		return nil, err
+	}
+
+	// verify the Hash_Envelope object
+	if err := message.Verify(nil, verifier); err != nil {
+		return nil, err
+	}
+
+	// cast to type Algorithm
+	hashAlgorithm, err := message.Headers.Protected.PayloadHashAlgorithm()
+	if err != nil {
+		return nil, err
+	}
+	message.Headers.Protected[HeaderLabelPayloadHashAlgorithm] = hashAlgorithm
+
+	// validate the hash value
+	if err := validateHash(hashAlgorithm, message.Payload); err != nil {
+		return nil, err
+	}
+
+	return &message, nil
+}
+
+// validateHash checks the validity of the known hash.
+func validateHash(alg Algorithm, value []byte) error {
+	hash := alg.hashFunc()
+	if hash == 0 {
+		return nil // no check on unsupported hash algorithms
+	}
+	if size := hash.Size(); size != len(value) {
+		return fmt.Errorf("%v: size mismatch: expected %d, got %d", alg, size, len(value))
+	}
+	return nil
+}
+
+// setHashEnvelopeProtectedHeader sets the protected header for a Hash_Envelope
+// object.
+func setHashEnvelopeProtectedHeader(base ProtectedHeader, payload *HashEnvelopePayload) ProtectedHeader {
+	header := maps.Clone(base)
+	if header == nil {
+		header = make(ProtectedHeader)
+	}
+	header[HeaderLabelPayloadHashAlgorithm] = payload.HashAlgorithm
+	if payload.PreimageContentType != nil {
+		header[HeaderLabelPayloadPreimageContentType] = payload.PreimageContentType
+	}
+	if payload.Location != "" {
+		header[HeaderLabelPayloadLocation] = payload.Location
+	}
+	return header
+}
+
+// validateHashEnvelopeHeaders validates the headers of a Hash_Envelope object.
+// See https://www.ietf.org/archive/id/draft-ietf-cose-hash-envelope-04.html
+// section 4 for more details.
+func validateHashEnvelopeHeaders(headers *Headers) error {
+	var foundPayloadHashAlgorithm bool
+	for label, value := range headers.Protected {
+		// Validate that all header labels are integers or strings.
+		// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-1.4
+		label, ok := normalizeLabel(label)
+		if !ok {
+			return errors.New("header label: require int / tstr type")
+		}
+
+		switch label {
+		case HeaderLabelContentType:
+			return errors.New("protected header parameter: content type: not allowed")
+		case HeaderLabelPayloadHashAlgorithm:
+			_, isAlg := value.(Algorithm)
+			if !isAlg && !canInt(value) {
+				return errors.New("protected header parameter: payload hash alg: require int type")
+			}
+			foundPayloadHashAlgorithm = true
+		case HeaderLabelPayloadPreimageContentType:
```

Comment:
What happens if `PreimageContentType` exists in both the protected and unprotected headers? Is it a valid case? If it is valid, what happens if they have different values? This scenario is not defined in the draft.

Reply:
Thanks @JeyJeyGao,
This is a good question that I've opened https://github.com/cose-wg/draft-ietf-cose-hash-envelope/issues/49

For future reference, you can always ask the question to the associated IETF discussion group. In this case: 
<a href="mailto:cose@ietf.org?subject=draft-ietf-cose-hash-envelope">cose@ietf.org</a> with a subject that contains: `draft-ietf-cose-hash-envelope`

The mail archive can be viewed here: [https://mailarchive.ietf.org/arch/browse/cose?q=draft-ietf-cose-hash-envelope](https://mailarchive.ietf.org/arch/browse/cose/?q=draft-ietf-cose-hash-envelope)

Reply:
I've double checked the [RFC 9052 section 3](https://www.rfc-editor.org/rfc/rfc9052.html#name-header-parameters), it states
> Applications **SHOULD** verify that the same label does not occur in both the protected and unprotected header parameters. If the message is not rejected as malformed, attributes MUST be obtained from the protected bucket, and only if an attribute is not found in the protected bucket can that attribute be obtained from the unprotected bucket.

In other words, the same label can exist in both headers although it is not suggested. It really depends on the application. Therefore, I would keep the current implementation and leave the discussion to the hash envelope draft.

Reply:
While the end result of above is the same, I'd suggest holding to minimize churn while we sort out https://github.com/cose-wg/draft-ietf-cose-hash-envelope/pull/50
We're all here in Bangkok this week for IETF 122, so I'm sure we can solve this quickly.

Reply:
I'd limit this library to simple validation of int / tstr for the labels, and leave use of these parameters to applications that build on this library.

Reply:
We've updated the hash-envelope draft to clarify the protected/unprotected headers: https://datatracker.ietf.org/doc/draft-ietf-cose-hash-envelope/05/

Reply:
@SteveLasker I've just updated this PR to meet draft 05